### ISSUE TITLE
ADIOS2 building error

### ISSUE THREAD

**Issue by hnakahara79** [ECP-WarpX/WarpX#3707]
Hi everyone, 

During the installation process (https://warpx.readthedocs.io/en/latest/install/hpc/perlmutter.html#running), I successfully build "c-blosc" but while building "ADIOS2" I received an error message. Please see the screen-shot of the error below.

How to fix this problem.

Thanks and best regards.

<img width="1426" alt="adios2" src="https://user-images.githubusercontent.com/125529792/220627581-924c25db-85c7-4e7e-9b00-59882f55da05.png">

**Comment by ax3l** [comment]
Hi @hnakahara79,

did you already `source $HOME/perlmutter_warpx.profile` before starting to install the dependencies as described here?
- https://warpx.readthedocs.io/en/latest/install/hpc.html#warpx-profile
- https://warpx.readthedocs.io/en/latest/install/hpc/perlmutter.html#running

If you don't do that, then:
- the wrong compilers will be used for your dependendencies (I see a wrong GCC, in a very old version, in your log showing up), and
- dependencies like c-blosc will not be found (because `export CMAKE_PREFIX_PATH=...` lines were not executed.)

Please redo the c-blosc installation after `source`ing that profile. Let me know if that fixes it.

**Comment by hnakahara79** [comment]
Hi @ax3l , thank you for pulling me back on the right path !!

I follow your suggestions. 

Now, when I do: "source $HOME/perlmutter_warpx.profile" 

one error message is appear for the line "export CUDACXX=$(which nvcc)" 

---> no nvcc

please, can you suggest how to fix this ?

Thank you.

**Comment by hnakahara79** [comment]
Hi @ax3l , do you think I have add the line: "module load nvidia/22.9" 

because if do this on the terminal then I can see "which nvcc" output.

**Comment by hnakahara79** [comment]
Hi @ax3l , now up to " source $HOME/perlmutter_warpx.profile " things looks fine. 

I am proceeding to install ADIOS2, BLAS++ and LAPACK++

c-blosc (I/O compression): building  **done cleanly**

but for "ADIOS2"  **another error appear**  (please see the screenshot below)

Although when I type **echo $CXX** on the command terminal **it gives me CC**

I don't understand where the problem is ??

<img width="1439" alt="adios_2" src="https://user-images.githubusercontent.com/125529792/220812388-8f77b72e-b956-4cf5-9908-7905f461a1b6.png">

**Comment by ax3l** [comment]
You want to update the last block of environment variables, CXX, CC, etc. to your compilers.

For instance,
```bash
export CC=$(which gcc)
export CXX=$(which g++)
export FC=$(which gfortran)
export CUDACXX=$(which nvcc)
export CUDAHOSTCXX=$(which g++)
```

What compiler module did you load?

**Comment by hnakahara79** [comment]
I loaded as given on the perlmutter_warpx.profile 

export CC=cc
export CXX=CC
export FC=ftn
export CUDACXX=$(which nvcc)
export CUDAHOSTCXX=CC

**Comment by ax3l** [comment]
Please change that block as quoted above. That block only applies for Cray/HPE machines and needs to be adopted for your machine.

For CUDA, please load the module cuda/11.7 and not the module nvidia/22.9. the former is the cuda toolkit, the later the Nvidia hpc toolkit. We work with both, but the former is more stable.

A host compiler, load the gnu/9.5.0 module 
https://github.com/ECP-WarpX/WarpX/issues/3695#issuecomment-1435515034

**Comment by hnakahara79** [comment]
OK, thank you. I will follow as you suggested and will let you know if there is any problem.

thanks also for gnu/9.5.0 suggestion.

**Comment by hnakahara79** [comment]
Hi @ax3l , now I could proceed further, but unfortunately 

cmake --build src/adios2-pm-build --target install -j 16

gives me an error:    **gmake: *** [all] Error 2**

**Comment by ax3l** [comment]
Hi @hnakahara79,

I suggest your `warpx.profile` (not `perlmutter_...`, you are not on perlmutter at NERSC right now) is structures like this:
```bash
# required dependencies
export PATH=/home/center/app/cmake-3.20.2/bin:${PATH}
module load gnu/9.5.0
module load cuda/11.7
module load openmpi/4.1.4

# optional dependencies (todo: later)
# TODO: boost for QED physics
# TODO: hdf5 for alternative openPMD output - ADIOS2 is fine
# TODO: BLAS++ & LAPACK++: for PSATD solver in RZ geometry

export CMAKE_PREFIX_PATH=$HOME/sw/c-blosc-1.21.1:${CMAKE_PREFIX_PATH}
export CMAKE_PREFIX_PATH=$HOME/sw/adios2-2.8.3:${CMAKE_PREFIX_PATH}

# optional dependency: Python (later)
# TODO: Python for PICMI Python interface

export AMREX_CUDA_ARCH=7.0

# compiler environment hints
export CC=$(which gcc)
export CXX=$(which g++)
export FC=$(which gfortran)
export CUDACXX=$(which nvcc)
export CUDAHOSTCXX=$(which g++)
```

**Comment by ax3l** [comment]
Open a new terminal and log again into the cluster.
After every login, as a first step, `source ~/warpx.profile`.

Now, please re-do the `c-blosc` and `adios2` installs. Please report the CMake errors that you get, if any.

```bash
cd $HOME
rm -rf src/c-blosc src/adios2 sw

# c-blosc (I/O compression)
git clone -b v1.21.1 https://github.com/Blosc/c-blosc.git src/c-blosc
rm -rf src/c-blosc-build
cmake -S src/c-blosc -B src/c-blosc-pm-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/c-blosc-1.21.1
cmake --build src/c-blosc-build --target install --parallel 16

# ADIOS2
git clone -b v2.8.3 https://github.com/ornladios/ADIOS2.git src/adios2
rm -rf src/adios2-build
cmake -S src/adios2 -B src/adios2-build -DADIOS2_USE_Blosc=ON -DADIOS2_BUILD_EXAMPLES=OFF -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DADIOS2_USE_ZeroMQ=OFF -DCMAKE_INSTALL_PREFIX=$HOME/sw/adios2-2.8.3
cmake --build src/adios2-build --target install -j 16
```

**Comment by hnakahara79** [comment]
Hi @ax3l , sorry for my mistakes. I was just following the perlmutter_warpx.profile for my own. 

Now, I will start again and will report to you if there is any problem. 

Sorry for the inconvenience and thanks for bearing with me.

**Comment by ax3l** [comment]
No worries, you are doing fine! We just need to adjust to your cluster :)

**Comment by hnakahara79** [comment]
Hi @ax3l , I follow your advice exactly. Building of blosc was fine, **but there was still an error with ADIOS2.** 

Please, see the files below for the details of the building result:

**Comment by ax3l** [comment]
@guj can you please triage this with upstream? The hdf5 examples in adios2 v2.8.3 have a linker error (please see the last comment), likely because some dependency is not forwarded properly.

@hnakahara79 please add these option to your ADIOS2 cake configuration line to continue: `-DADIOS2_BUILD_EXAMPLES=OFF`

**Comment by hnakahara79** [comment]
Hi @ax3l thank you for the suggestion. I will try this today.

**Comment by hnakahara79** [comment]
Hi @ax3l thank you for your continuous help, now ADIOS2 is also build.

Do you think now I can proceed further to build # BLAS++ (for PSATD+RZ)  and  # LAPACK++ (for PSATD+RZ)  ?

**Comment by guj** [comment]
> 

Here is what I think:

First, if one plans to use openPMD-api, it is not necessary to compile ADIOS with HDF5 (i.e. can add the o option -DADIOS2_USE_HDF5=OFF to cmake).

Second, to compile with HDF5 support in ADIOS, one needs to make sure to use the same mpi libraries for both  HDF5  and ADIOS in order to avoid symbol  errors.  Both Norbert and I think the error messages shown points to conflict in the mpi libraries.

- Junmin

**Comment by hnakahara79** [comment]
Hi @guj and @ax3l thank you for explaining in a bit detail and reminding about the caution that must be taken during compilation.

**Comment by ax3l** [comment]
> Do you think now I can proceed further to build # BLAS++ (for PSATD+RZ) and # LAPACK++ (for PSATD+RZ) ?

Yes, please disable HDF5 for ADIOS2 as Junmin suggested (we don't use it) and continue as planned.

BLAS++ and LAPACK++ is only needed for WarpX in RZ geometry with PSATD field solver.

If you like, you can now test the executable you got so far with our 3D input sets (or 2D or 1D).

**Comment by hnakahara79** [comment]
Hi @ax3l , thank you for the explanation.

At the moment I am looking forward to work with 3D cartesian geometry case, so I will postpone building of BLAS++ and LAPACK++ for future.

Before testing the simulation, I have to compile the code **(please correct me if I am wrong)** as described in https://warpx.readthedocs.io/en/latest/install/cmake.html#install-developers. I will follow that part of the documentation. 

I also want to use GPU for running the code. I saw that there is a separate documentation for it https://amrex-codes.github.io/amrex/docs_html/GPU.html#building-gpu-support. But if I have any difficulty on running with GPU, I will consult you. 

Thank you.

**Comment by hnakahara79** [comment]
Hi @ax3l , finally I am able to generate the executables :)

Thanks a lot for your kindness and help.

Best regards.
